File: museum_3d.py
// SPDX-License-Identifier: MIT
# SPDX-License-Identifier: MIT

import sqlite3
import json
from flask import Flask, render_template_string, jsonify, request
from datetime import datetime

app = Flask(__name__)
DB_PATH = 'rustchain.db'

@app.route('/museum')
def museum_3d():
    return render_template_string('''
<!DOCTYPE html>
<html>
<head>
    <title>3D Vintage Hardware Museum</title>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/three.js/r128/three.min.js"></script>
    <style>
        body { margin: 0; overflow: hidden; background: #000; }
        canvas { display: block; }
        #info { position: absolute; top: 10px; left: 10px; color: white; font-family: Arial; z-index: 100; }
        #controls { position: absolute; bottom: 10px; left: 10px; color: white; font-family: Arial; font-size: 12px; }
        #stats-panel { position: absolute; top: 10px; right: 10px; width: 300px; background: rgba(0,0,0,0.8); color: white; padding: 15px; border-radius: 5px; display: none; }
        .loading { position: absolute; top: 50%; left: 50%; transform: translate(-50%, -50%); color: white; font-family: Arial; }
    </style>
</head>
<body>
    <div id="info">
        <h3>RustChain 3D Hardware Museum</h3>
        <p>Click on exhibits to view mining stats</p>
    </div>
    <div id="controls">
        WASD: Move | Mouse: Look | Click: Interact | Mobile: Touch to move
    </div>
    <div id="stats-panel">
        <h4 id="exhibit-name">Hardware Info</h4>
        <div id="exhibit-stats"></div>
        <button onclick="closeStats()" style="margin-top: 10px; padding: 5px 10px;">Close</button>
    </div>
    <div class="loading" id="loading">Loading Museum...</div>

    <script>
        let scene, camera, renderer, controls;
        let exhibits = [];
        let moveForward = false, moveBackward = false, moveLeft = false, moveRight = false;
        let velocity = new THREE.Vector3();
        let direction = new THREE.Vector3();
        let mouse = new THREE.Vector2();
        let raycaster = new THREE.Raycaster();
        let touchStartX = 0, touchStartY = 0;

        const exhibitData = {
            'powerbook_g4': {
                name: 'PowerBook G4 (2005)',
                position: [-5, 0, 0],
                color: 0x808080,
                api_endpoint: 'powerbook_g4'
            },
            'ibm_power8': {
                name: 'IBM POWER8 Server',
                position: [5, 0, 0], 
                color: 0x000080,
                api_endpoint: 'ibm_power8'
            },
            'sun_sparc': {
                name: 'Sun SPARC Station',
                position: [0, 0, -5],
                color: 0xff6600,
                api_endpoint: 'sun_sparc'
            },
            'sgi_octane': {
                name: 'SGI Octane',
                position: [-3, 0, 5],
                color: 0x4b0082,
                api_endpoint: 'sgi_octane'
            },
            'dec_alpha': {
                name: 'DEC Alpha Workstation',
                position: [3, 0, 5],
                color: 0x800000,
                api_endpoint: 'dec_alpha'
            }
        };

        function init() {
            scene = new THREE.Scene();
            scene.fog = new THREE.Fog(0x000000, 0, 750);

            camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 1, 1000);
            camera.position.set(0, 5, 10);

            renderer = new THREE.WebGLRenderer();
            renderer.setSize(window.innerWidth, window.innerHeight);
            renderer.setClearColor(0x000000);
            document.body.appendChild(renderer.domElement);

            // Lighting
            const ambientLight = new THREE.AmbientLight(0x404040, 0.4);
            scene.add(ambientLight);

            const directionalLight = new THREE.DirectionalLight(0xffffff, 0.8);
            directionalLight.position.set(10, 10, 5);
            scene.add(directionalLight);

            // Floor
            const floorGeometry = new THREE.PlaneGeometry(100, 100);
            const floorMaterial = new THREE.MeshLambertMaterial({ color: 0x333333 });
            const floor = new THREE.Mesh(floorGeometry, floorMaterial);
            floor.rotation.x = -Math.PI / 2;
            scene.add(floor);

            // Grid
            const gridHelper = new THREE.GridHelper(100, 100, 0x444444, 0x222222);
            scene.add(gridHelper);

            // Create exhibits
            Object.keys(exhibitData).forEach(key => {
                const data = exhibitData[key];
                const geometry = new THREE.BoxGeometry(2, 1.5, 1);
                const material = new THREE.MeshLambertMaterial({ color: data.color });
                const exhibit = new THREE.Mesh(geometry, material);
                exhibit.position.set(...data.position);
                exhibit.position.y = 0.75;
                exhibit.userData = { type: 'exhibit', id: key, name: data.name };
                
                // Add glow effect
                const glowGeometry = new THREE.BoxGeometry(2.2, 1.7, 1.2);
                const glowMaterial = new THREE.MeshBasicMaterial({
                    color: data.color,
                    transparent: true,
                    opacity: 0.3
                });
                const glow = new THREE.Mesh(glowGeometry, glowMaterial);
                glow.position.copy(exhibit.position);
                scene.add(glow);
                
                scene.add(exhibit);
                exhibits.push(exhibit);

                // Label
                const loader = new THREE.FontLoader();
                createTextLabel(data.name, exhibit.position.clone().add(new THREE.Vector3(0, 2, 0)));
            });

            // Event listeners
            document.addEventListener('keydown', onKeyDown);
            document.addEventListener('keyup', onKeyUp);
            document.addEventListener('click', onMouseClick);
            document.addEventListener('mousemove', onMouseMove);
            
            // Touch events for mobile
            document.addEventListener('touchstart', onTouchStart, { passive: false });
            document.addEventListener('touchmove', onTouchMove, { passive: false });
            document.addEventListener('touchend', onTouchEnd, { passive: false });

            document.getElementById('loading').style.display = 'none';
        }

        function createTextLabel(text, position) {
            const canvas = document.createElement('canvas');
            const context = canvas.getContext('2d');
            context.font = '48px Arial';
            context.fillStyle = 'white';
            context.fillText(text, 0, 48);

            const texture = new THREE.CanvasTexture(canvas);
            const spriteMaterial = new THREE.SpriteMaterial({ map: texture });
            const sprite = new THREE.Sprite(spriteMaterial);
            sprite.position.copy(position);
            sprite.scale.set(4, 2, 1);
            scene.add(sprite);
        }

        function onKeyDown(event) {
            switch(event.code) {
                case 'KeyW': moveForward = true; break;
                case 'KeyS': moveBackward = true; break;
                case 'KeyA': moveLeft = true; break;
                case 'KeyD': moveRight = true; break;
            }
        }

        function onKeyUp(event) {
            switch(event.code) {
                case 'KeyW': moveForward = false; break;
                case 'KeyS': moveBackward = false; break;
                case 'KeyA': moveLeft = false; break;
                case 'KeyD': moveRight = false; break;
            }
        }

        function onMouseClick(event) {
            mouse.x = (event.clientX / window.innerWidth) * 2 - 1;
            mouse.y = -(event.clientY / window.innerHeight) * 2 + 1;

            raycaster.setFromCamera(mouse, camera);
            const intersects = raycaster.intersectObjects(exhibits);

            if (intersects.length > 0) {
                const clicked = intersects[0].object;
                if (clicked.userData.type === 'exhibit') {
                    showExhibitStats(clicked.userData.id, clicked.userData.name);
                }
            }
        }

        function onMouseMove(event) {
            const movementX = event.movementX || 0;
            const movementY = event.movementY || 0;

            camera.rotation.y -= movementX * 0.002;
            camera.rotation.x -= movementY * 0.002;
            camera.rotation.x = Math.max(-Math.PI/2, Math.min(Math.PI/2, camera.rotation.x));
        }

        function onTouchStart(event) {
            event.preventDefault();
            touchStartX = event.touches[0].clientX;
            touchStartY = event.touches[0].clientY;
        }

        function onTouchMove(event) {
            event.preventDefault();
            if (event.touches.length === 1) {
                const deltaX = event.touches[0].clientX - touchStartX;
                const deltaY = event.touches[0].clientY - touchStartY;
                
                camera.rotation.y -= deltaX * 0.005;
                camera.rotation.x -= deltaY * 0.005;
                camera.rotation.x = Math.max(-Math.PI/2, Math.min(Math.PI/2, camera.rotation.x));
                
                touchStartX = event.touches[0].clientX;
                touchStartY = event.touches[0].clientY;
            }
        }

        function onTouchEnd(event) {
            event.preventDefault();
            if (event.changedTouches.length === 1) {
                mouse.x = (event.changedTouches[0].clientX / window.innerWidth) * 2 - 1;
                mouse.y = -(event.changedTouches[0].clientY / window.innerHeight) * 2 + 1;
                
                raycaster.setFromCamera(mouse, camera);
                const intersects = raycaster.intersectObjects(exhibits);
                
                if (intersects.length > 0) {
                    const clicked = intersects[0].object;
                    if (clicked.userData.type === 'exhibit') {
                        showExhibitStats(clicked.userData.id, clicked.userData.name);
                    }
                }
            }
        }

        function showExhibitStats(exhibitId, exhibitName) {
            document.getElementById('exhibit-name').textContent = exhibitName;
            document.getElementById('stats-panel').style.display = 'block';
            document.getElementById('exhibit-stats').innerHTML = '<p>Loading stats...</p>';

            fetch(`/api/miner_stats/${exhibitId}`)
                .then(response => response.json())
                .then(data => {
                    let html = `
                        <p><strong>Hash Rate:</strong> ${data.hash_rate || 'N/A'}</p>
                        <p><strong>Temperature:</strong> ${data.temperature || 'N/A'}°C</p>
                        <p><strong>Power Usage:</strong> ${data.power_usage || 'N/A'}W</p>
                        <p><strong>Uptime:</strong> ${data.uptime || 'N/A'}</p>
                        <p><strong>Blocks Mined:</strong> ${data.blocks_mined || 0}</p>
                        <p><strong>Status:</strong> <span style="color: ${data.status === 'active' ? 'green' : 'red'}">${data.status || 'Unknown'}</span></p>
                    `;
                    if (data.last_block_time) {
                        html += `<p><strong>Last Block:</strong> ${new Date(data.last_block_time).toLocaleString()}</p>`;
                    }
                    document.getElementById('exhibit-stats').innerHTML = html;
                })
                .catch(error => {
                    document.getElementById('exhibit-stats').innerHTML = '<p>Error loading stats</p>';
                });
        }

        function closeStats() {
            document.getElementById('stats-panel').style.display = 'none';
        }

        function animate() {
            requestAnimationFrame(animate);

            direction.x = Number(moveRight) - Number(moveLeft);
            direction.z = Number(moveBackward) - Number(moveForward);
            direction.normalize();

            if (moveForward || moveBackward) velocity.z -= direction.z * 400.0 * 0.016;
            if (moveLeft || moveRight) velocity.x -= direction.x * 400.0 * 0.016;

            velocity.x *= 0.9;
            velocity.z *= 0.9;

            camera.translateX(velocity.x * 0.016);
            camera.translateZ(velocity.z * 0.016);

            if (camera.position.y < 2) camera.position.y = 2;

            renderer.render(scene, camera);
        }

        function onWindowResize() {
            camera.aspect = window.innerWidth / window.innerHeight;
            camera.updateProjectionMatrix();
            renderer.setSize(window.innerWidth, window.innerHeight);
        }

        window.addEventListener('resize', onWindowResize);
        window.addEventListener('click', () => {
            document.body.requestPointerLock();
        });

        init();
        animate();
    </script>
</body>
</html>
    ''')

@app.route('/api/miner_stats/<exhibit_id>')
def miner_stats_api(exhibit_id):
    try:
        with sqlite3.connect(DB_PATH) as conn:
            cursor = conn.cursor()
            
            # Get miner stats for exhibit
            cursor.execute('''
                SELECT hash_rate, temperature, power_usage, uptime, blocks_mined, status, last_block_time
                FROM vintage_miners 
                WHERE exhibit_id = ?
            ''', (exhibit_id,))
            
            result = cursor.fetchone()
            
            if result:
                return jsonify({
                    'hash_rate': f"{result[0]} MH/s" if result[0] else None,
                    'temperature': result[1],
                    'power_usage': result[2],
                    'uptime': f"{result[3]} hours" if result[3] else None,
                    'blocks_mined': result[4],
                    'status': result[5],
                    'last_block_time': result[6]
                })
            else:
                # Return mock data for demo
                mock_data = {
                    'powerbook_g4': {
                        'hash_rate': '12.5 KH/s',
                        'temperature': 68,
                        'power_usage': 85,
                        'uptime': '72',
                        'blocks_mined': 3,
                        'status': 'active',
                        'last_block_time': datetime.now().isoformat()
                    },
                    'ibm_power8': {
                        'hash_rate': '2.1 GH/s', 
                        'temperature': 45,
                        'power_usage': 350,
                        'uptime': '168',
                        'blocks_mined': 47,
                        'status': 'active',
                        'last_block_time': datetime.now().isoformat()
                    },
                    'sun_sparc': {
                        'hash_rate': '850 KH/s',
                        'temperature': 52,
                        'power_usage': 120,
                        'uptime': '96',
                        'blocks_mined': 12,
                        'status': 'maintenance',
                        'last_block_time': datetime.now().isoformat()
                    },
                    'sgi_octane': {
                        'hash_rate': '1.2 MH/s',
                        'temperature': 41,
                        'power_usage': 200,
                        'uptime': '240',
                        'blocks_mined': 28,
                        'status': 'active',
                        'last_block_time': datetime.now().isoformat()
                    },
                    'dec_alpha': {
                        'hash_rate': '950 KH/s',
                        'temperature': 39,
                        'power_usage': 175,
                        'uptime': '144',
                        'blocks_mined': 19,
                        'status': 'offline',
                        'last_block_time': None
                    }
                }
                
                return jsonify(mock_data.get(exhibit_id, {
                    'hash_rate': None,
                    'temperature': None,
                    'power_usage': None,
                    'uptime': None,
                    'blocks_mined': 0,
                    'status': 'unknown',
                    'last_block_time': None
                }))
                
    except Exception as e:
        return jsonify({'error': str(e)}), 500

def init_museum_db():
    try:
        with sqlite3.connect(DB_PATH) as conn:
            cursor = conn.cursor()
            cursor.execute('''
                CREATE TABLE IF NOT EXISTS vintage_miners (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    exhibit_id TEXT UNIQUE,
                    name TEXT,
                    hash_rate REAL,
                    temperature INTEGER,
                    power_usage INTEGER,
                    uptime INTEGER,
                    blocks_mined INTEGER DEFAULT 0,
                    status TEXT DEFAULT 'offline',
                    last_block_time TEXT,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                )
            ''')
            conn.commit()
    except Exception as e:
        print(f"Database init error: {e}")

if __name__ == '__main__':
    init_museum_db()
    app.run(debug=True)

File: static/js/museum3d.js
// SPDX-License-Identifier: MIT

class Museum3D {
    constructor() {
        this.scene = null;
        this.camera = null;
        this.renderer = null;
        this.controls = null;
        this.exhibits = [];
        this.raycaster = new THREE.Raycaster();
        this.mouse = new THREE.Vector2();
        this.keys = {};
        this.moveSpeed = 0.1;
        this.isMobile = /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
        this.touchStart = { x: 0, y: 0 };
        this.touchEnd = { x: 0, y: 0 };
        this.lastApiUpdate = 0;
        this.apiUpdateInterval = 30000; // 30 seconds
        
        this.init();
        this.animate();
        this.setupEventListeners();
        this.loadExhibits();
    }

    init() {
        // Scene setup
        this.scene = new THREE.Scene();
        this.scene.background = new THREE.Color(0x0a0a0a);
        this.scene.fog = new THREE.Fog(0x0a0a0a, 10, 100);

        // Camera setup
        this.camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000);
        this.camera.position.set(0, 2, 5);

        // Renderer setup
        this.renderer = new THREE.WebGLRenderer({ antialias: true });
        this.renderer.setSize(window.innerWidth, window.innerHeight);
        this.renderer.shadowMap.enabled = true;
        this.renderer.shadowMap.type = THREE.PCFSoftShadowMap;
        this.renderer.toneMapping = THREE.ACESFilmicToneMapping;
        this.renderer.toneMappingExposure = 1.2;
        document.getElementById('museum-container').appendChild(this.renderer.domElement);

        // Lighting
        this.setupLighting();
        
        // Floor
        this.createFloor();
        
        // Create exhibit pedestals
        this.createPedestals();
    }

    setupLighting() {
        // Ambient light
        const ambientLight = new THREE.AmbientLight(0x404040, 0.3);
        this.scene.add(ambientLight);

        // Main directional light
        const directionalLight = new THREE.DirectionalLight(0xffffff, 0.8);
        directionalLight.position.set(10, 10, 5);
        directionalLight.castShadow = true;
        directionalLight.shadow.mapSize.width = 2048;
        directionalLight.shadow.mapSize.height = 2048;
        directionalLight.shadow.camera.near = 0.5;
        directionalLight.shadow.camera.far = 50;
        this.scene.add(directionalLight);

        // Exhibit spot lights
        const spotLight1 = new THREE.SpotLight(0xffffff, 1, 10, Math.PI * 0.3, 0.3, 1);
        spotLight1.position.set(-3, 5, 0);
        spotLight1.target.position.set(-3, 0, 0);
        this.scene.add(spotLight1);
        this.scene.add(spotLight1.target);

        const spotLight2 = new THREE.SpotLight(0xffffff, 1, 10, Math.PI * 0.3, 0.3, 1);
        spotLight2.position.set(3, 5, 0);
        spotLight2.target.position.set(3, 0, 0);
        this.scene.add(spotLight2);
        this.scene.add(spotLight2.target);
    }

    createFloor() {
        const floorGeometry = new THREE.PlaneGeometry(50, 50);
        const floorMaterial = new THREE.MeshLambertMaterial({ 
            color: 0x222222,
            transparent: true,
            opacity: 0.8
        });
        const floor = new THREE.Mesh(floorGeometry, floorMaterial);
        floor.rotation.x = -Math.PI / 2;
        floor.receiveShadow = true;
        this.scene.add(floor);
    }

    createPedestals() {
        const pedestalGeometry = new THREE.CylinderGeometry(1.2, 1.2, 0.3, 8);
        const pedestalMaterial = new THREE.MeshPhongMaterial({ color: 0x333333 });
        
        // PowerBook G4 pedestal
        const pedestal1 = new THREE.Mesh(pedestalGeometry, pedestalMaterial);
        pedestal1.position.set(-3, 0.15, 0);
        pedestal1.castShadow = true;
        pedestal1.receiveShadow = true;
        this.scene.add(pedestal1);

        // IBM POWER8 pedestal
        const pedestal2 = new THREE.Mesh(pedestalGeometry, pedestalMaterial);
        pedestal2.position.set(3, 0.15, 0);
        pedestal2.castShadow = true;
        pedestal2.receiveShadow = true;
        this.scene.add(pedestal2);
    }

    loadExhibits() {
        this.createPowerBookG4();
        this.createIBMPOWER8();
        this.updateExhibitData();
    }

    createPowerBookG4() {
        const group = new THREE.Group();
        
        // Laptop base
        const baseGeometry = new THREE.BoxGeometry(1.6, 0.1, 1.2);
        const baseMaterial = new THREE.MeshPhongMaterial({ color: 0xc0c0c0 });
        const base = new THREE.Mesh(baseGeometry, baseMaterial);
        base.castShadow = true;
        group.add(base);

        // Screen
        const screenGeometry = new THREE.BoxGeometry(1.5, 1.0, 0.05);
        const screenMaterial = new THREE.MeshPhongMaterial({ color: 0x1a1a1a });
        const screen = new THREE.Mesh(screenGeometry, screenMaterial);
        screen.position.set(0, 0.55, -0.5);
        screen.rotation.x = -Math.PI * 0.1;
        screen.castShadow = true;
        group.add(screen);

        // Logo
        const logoGeometry = new THREE.CircleGeometry(0.08, 16);
        const logoMaterial = new THREE.MeshPhongMaterial({ color: 0xffffff });
        const logo = new THREE.Mesh(logoGeometry, logoMaterial);
        logo.position.set(0, 0.8, -0.47);
        logo.rotation.x = -Math.PI * 0.1;
        group.add(logo);

        group.position.set(-3, 0.35, 0);
        group.userData = { 
            type: 'powerbook_g4',
            name: 'PowerBook G4',
            specs: 'PowerPC G4 • 1.67GHz • 2GB RAM',
            minerId: 'pbg4_001'
        };
        
        this.scene.add(group);
        this.exhibits.push(group);
    }

    createIBMPOWER8() {
        const group = new THREE.Group();
        
        // Server chassis
        const chassisGeometry = new THREE.BoxGeometry(1.8, 0.8, 2.0);
        const chassisMaterial = new THREE.MeshPhongMaterial({ color: 0x2a2a2a });
        const chassis = new THREE.Mesh(chassisGeometry, chassisMaterial);
        chassis.castShadow = true;
        group.add(chassis);

        // Front panel
        const frontGeometry = new THREE.BoxGeometry(1.7, 0.7, 0.1);
        const frontMaterial = new THREE.MeshPhongMaterial({ color: 0x1a1a1a });
        const front = new THREE.Mesh(frontGeometry, frontMaterial);
        front.position.set(0, 0, 1.05);
        group.add(front);

        // LED indicators
        const ledGeometry = new THREE.SphereGeometry(0.02, 8, 8);
        const ledMaterial = new THREE.MeshPhongMaterial({ 
            color: 0x00ff00,
            emissive: 0x004400
        });
        
        for (let i = 0; i < 4; i++) {
            const led = new THREE.Mesh(ledGeometry, ledMaterial);
            led.position.set(-0.6 + i * 0.4, 0.2, 1.1);
            group.add(led);
        }

        // IBM logo
        const logoGeometry = new THREE.PlaneGeometry(0.3, 0.1);
        const logoMaterial = new THREE.MeshPhongMaterial({ color: 0x0066cc });
        const logo = new THREE.Mesh(logoGeometry, logoMaterial);
        logo.position.set(0, -0.2, 1.1);
        group.add(logo);

        group.position.set(3, 0.7, 0);
        group.userData = { 
            type: 'ibm_power8',
            name: 'IBM POWER8',
            specs: 'POWER8 • 8-Core • 64GB RAM',
            minerId: 'power8_001'
        };
        
        this.scene.add(group);
        this.exhibits.push(group);
    }

    setupEventListeners() {
        // Keyboard controls
        document.addEventListener('keydown', (event) => {
            this.keys[event.code] = true;
        });
        
        document.addEventListener('keyup', (event) => {
            this.keys[event.code] = false;
        });

        // Mouse/touch controls
        if (this.isMobile) {
            this.setupTouchControls();
        } else {
            this.setupMouseControls();
        }

        // Window resize
        window.addEventListener('resize', () => {
            this.camera.aspect = window.innerWidth / window.innerHeight;
            this.camera.updateProjectionMatrix();
            this.renderer.setSize(window.innerWidth, window.innerHeight);
        });

        // Click/tap interaction
        this.renderer.domElement.addEventListener('click', (event) => {
            this.handleClick(event);
        });
    }

    setupMouseControls() {
        let isMouseDown = false;
        let previousMousePosition = { x: 0, y: 0 };

        this.renderer.domElement.addEventListener('mousedown', (event) => {
            isMouseDown = true;
            previousMousePosition = { x: event.clientX, y: event.clientY };
        });

        this.renderer.domElement.addEventListener('mouseup', () => {
            isMouseDown = false;
        });

        this.renderer.domElement.addEventListener('mousemove', (event) => {
            if (isMouseDown) {
                const deltaMove = {
                    x: event.clientX - previousMousePosition.x,
                    y: event.clientY - previousMousePosition.y
                };

                this.camera.rotation.y -= deltaMove.x * 0.005;
                this.camera.rotation.x -= deltaMove.y * 0.005;
                this.camera.rotation.x = Math.max(-Math.PI/2, Math.min(Math.PI/2, this.camera.rotation.x));

                previousMousePosition = { x: event.clientX, y: event.clientY };
            }
        });
    }

    setupTouchControls() {
        this.renderer.domElement.addEventListener('touchstart', (event) => {
            if (event.touches.length === 1) {
                this.touchStart = {
                    x: event.touches[0].clientX,
                    y: event.touches[0].clientY
                };
            }
        });

        this.renderer.domElement.addEventListener('touchmove', (event) => {
            event.preventDefault();
            if (event.touches.length === 1) {
                const deltaMove = {
                    x: event.touches[0].clientX - this.touchStart.x,
                    y: event.touches[0].clientY - this.touchStart.y
                };

                this.camera.rotation.y -= deltaMove.x * 0.005;
                this.camera.rotation.x -= deltaMove.y * 0.005;
                this.camera.rotation.x = Math.max(-Math.PI/2, Math.min(Math.PI/2, this.camera.rotation.x));

                this.touchStart = {
                    x: event.touches[0].clientX,
                    y: event.touches[0].clientY
                };
            }
        });

        // Virtual D-pad for mobile
        this.createVirtualControls();
    }

    createVirtualControls() {
        const controlsDiv = document.createElement('div');
        controlsDiv.id = 'virtual-controls';
        controlsDiv.style.cssText = `
            position: fixed;
            bottom: 20px;
            left: 20px;
            display: grid;
            grid-template-columns: 1fr 1fr 1fr;
            grid-template-rows: 1fr 1fr 1fr;
            gap: 10px;
            width: 120px;
            height: 120px;
            z-index: 1000;
        `;

        const buttons = [
            { text: '↖', keys: ['KeyW', 'KeyA'] },
            { text: '↑', keys: ['KeyW'] },
            { text: '↗', keys: ['KeyW', 'KeyD'] },
            { text: '←', keys: ['KeyA'] },
            { text: '⊙', keys: [] },
            { text: '→', keys: ['KeyD'] },
            { text: '↙', keys: ['KeyS', 'KeyA'] },
            { text: '↓', keys: ['KeyS'] },
            { text: '↘', keys: ['KeyS', 'KeyD'] }
        ];

        buttons.forEach((btn, index) => {
            if (btn.keys.length === 0) return; // Skip center button
            
            const button = document.createElement('button');
            button.textContent = btn.text;
            button.style.cssText = `
                background: rgba(255,255,255,0.2);
                border: 1px solid rgba(255,255,255,0.3);
                color: white;
                font-size: 16px;
                border-radius: 5px;
                touch-action: manipulation;
            `;

            button.addEventListener('touchstart', (e) => {
                e.preventDefault();
                btn.keys.forEach(key => this.keys[key] = true);
            });

            button.addEventListener('touchend', (e) => {
                e.preventDefault();
                btn.keys.forEach(key => this.keys[key] = false);
            });

            controlsDiv.appendChild(button);
        });

        document.body.appendChild(controlsDiv);
    }

    handleClick(event) {
        this.mouse.x = (event.clientX / window.innerWidth) * 2 - 1;
        this.mouse.y = -(event.clientY / window.innerHeight) * 2 + 1;

        this.raycaster.setFromCamera(this.mouse, this.camera);
        const intersects = this.raycaster.intersectObjects(this.exhibits, true);

        if (intersects.length > 0) {
            const exhibit = intersects[0].object.parent || intersects[0].object;
            if (exhibit.userData.type) {
                this.showExhibitInfo(exhibit.userData);
            }
        }
    }

    showExhibitInfo(exhibitData) {
        // Remove existing info panel
        const existing = document.getElementById('exhibit-info');
        if (existing) existing.remove();

        const infoPanel = document.createElement('div');
        infoPanel.id = 'exhibit-info';
        infoPanel.style.cssText = `
            position: fixed;
            top: 50%;
            left: 50%;
            transform: translate(-50%, -50%);
            background: rgba(0, 0, 0, 0.9);
            color: white;
            padding: 30px;
            border-radius: 10px;
            border: 1px solid #333;
            max-width: 400px;
            z-index: 1000;
            font-family: 'Courier New', monospace;
        `;

        // Get live data for this exhibit
        this.fetchExhibitData(exhibitData.minerId).then(data => {
            infoPanel.innerHTML = `
                <h2 style="margin-top: 0; color: #00ff00;">${exhibitData.name}</h2>
                <p style="color: #ccc;">${exhibitData.specs}</p>
                <hr style="border-color: #333;">
                <h3>Mining Stats:</h3>
                <div style="font-family: monospace; font-size: 14px;">
                    <div>Hashrate: <span style="color: #00ff00;">${data.hashrate || 'N/A'}</span></div>
                    <div>Power Draw: <span style="color: #ffaa00;">${data.power || 'N/A'}</span></div>
                    <div>Temperature: <span style="color: #ff6600;">${data.temperature || 'N/A'}</span></div>
                    <div>Uptime: <span style="color: #0099ff;">${data.uptime || 'N/A'}</span></div>
                    <div>Blocks Found: <span style="color: #ff00ff;">${data.blocks || '0'}</span></div>
                </div>
                <button onclick="document.getElementById('exhibit-info').remove()" 
                        style="margin-top: 20px; padding: 10px 20px; background: #333; color: white; border: none; border-radius: 5px; cursor: pointer;">
                    Close
                </button>
            `;
        });

        document.body.appendChild(infoPanel);
    }

    async fetchExhibitData(minerId) {
        try {
            const response = await fetch(`/api/miner/${minerId}/stats`);
            if (response.ok) {
                return await response.json();
            }
        } catch (error) {
            console.warn('Failed to fetch exhibit data:', error);
        }
        
        // Return mock data if API fails
        return {
            hashrate: Math.floor(Math.random() * 1000) + 'H/s',
            power: Math.floor(Math.random() * 200) + 'W',
            temperature: Math.floor(Math.random() * 40 + 40) + '°C',
            uptime: Math.floor(Math.random() * 24) + 'h ' + Math.floor(Math.random() * 60) + 'm',
            blocks: Math.floor(Math.random() * 10)
        };
    }

    updateExhibitData() {
        const now = Date.now();
        if (now - this.lastApiUpdate > this.apiUpdateInterval) {
            this.exhibits.forEach(exhibit => {
                if (exhibit.userData.minerId) {
                    this.fetchExhibitData(exhibit.userData.minerId).then(data => {
                        exhibit.userData.liveData = data;
                    });
                }
            });
            this.lastApiUpdate = now;
        }
    }

    handleMovement() {
        const moveVector = new THREE.Vector3();
        
        if (this.keys['KeyW']) moveVector.z -= this.moveSpeed;
        if (this.keys['KeyS']) moveVector.z += this.moveSpeed;
        if (this.keys['KeyA']) moveVector.x -= this.moveSpeed;
        if (this.keys['KeyD']) moveVector.x += this.moveSpeed;

        // Apply camera rotation to movement
        moveVector.applyQuaternion(this.camera.quaternion);
        this.camera.position.add(moveVector);

        // Keep camera above ground
        this.camera.position.y = Math.max(1, this.camera.position.y);
    }

    animate() {
        requestAnimationFrame(() => this.animate());
        
        this.handleMovement();
        this.updateExhibitData();
        
        // Animate LEDs on IBM POWER8
        const time = Date.now() * 0.001;
        this.exhibits.forEach(exhibit => {
            if (exhibit.userData.type === 'ibm_power8') {
                exhibit.children.forEach(child => {
                    if (child.material && child.material.emissive) {
                        const intensity = Math.sin(time * 3 + child.position.x * 10) * 0.5 + 0.5;
                        child.material.emissive.setHex(intensity > 0.5 ? 0x004400 : 0x001100);
                    }
                });
            }
        });
        
        this.renderer.render(this.scene, this.camera);
    }
}

// Initialize museum when page loads
document.addEventListener('DOMContentLoaded', () => {
    if (document.getElementById('museum-container')) {
        new Museum3D();
    }
});